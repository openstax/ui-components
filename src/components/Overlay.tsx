import styled from "styled-components";
import { CloseModalButton } from "./CloseModalButton";
import { Mask, ModalWrapper } from "./Modal";
import * as RAC from "react-aria-components";
import React from "react";

export const OverlayMask = styled(Mask)`
  background-color: rgba(0, 0, 0, 0.89);
`;

export const OverlayCloseButton = styled(CloseModalButton)`
  height: 4rem;
  width: 4rem;
  position: absolute;
  right: 2em;
  top: 2em;
`;

export const OverlayWrapper = styled(ModalWrapper)`
  color: #fff;
`;

export const OverlayBody = styled(RAC.Dialog)`
  position: relative;
  flex-grow: 1;
  height: 100%;
  z-index: 1;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  outline: none;
`;

export const Overlay = ({
  className,
  onClose,
  children,
  show,
  ...props
}: React.PropsWithChildren<{
  onClose: () => void;
  className?: string;
  show?: boolean;
}> & RAC.ModalOverlayProps) => {
  if (!show) { return null; }
  return (
    <OverlayMask
      className={className}
      isDismissable
      onOpenChange={(isOpen) => (!isOpen && onClose())}
      {...props}
    >
      <OverlayWrapper defaultOpen={true}>
        <OverlayCloseButton onClick={onClose} variant={'inverted-circle'} />
        <OverlayBody>
          { children }
        </OverlayBody>
      </OverlayWrapper>
    </OverlayMask>
  );
};
