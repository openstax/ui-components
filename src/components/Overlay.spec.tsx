import { act, render } from '@testing-library/react';
import { Overlay } from './Overlay';
import userEvent from '@testing-library/user-event';

describe('Overlay', () => {
  let root: HTMLElement;

  beforeEach(() => {
    root = document.createElement('main');
    root.id = 'root';
    document.body.append(root);
  });

  it('matches snapshot', () => {
    render(
      <Overlay onClose={jest.fn()} show={true}>
        Inner content
      </Overlay>, { container: root }
    );
    expect(document.body).toMatchSnapshot();
  });

  it('hides', () => {
    render(
      <Overlay onClose={jest.fn()} show={false}>
        Inner content
      </Overlay>, { container: root }
    );
    expect(document.body).toMatchSnapshot();
  });

  it('calls onClose when close button is clicked', async () => {
    const onClose = jest.fn();
    const user = userEvent.setup();

    await act(async() => {
      const overlay = render(<Overlay show onClose={onClose} />);
      await user.type(overlay.container, '{esc}');
    });

    expect(onClose).toHaveBeenCalled();
  });
});
