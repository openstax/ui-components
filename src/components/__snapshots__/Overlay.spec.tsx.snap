// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overlay hides 1`] = `
<body>
  <main
    id="root"
  />
</body>
`;

exports[`Overlay matches snapshot 1`] = `
<body>
  <main
    id="root"
  />
  <div
    class="sc-iBkjds sc-iqcoie cbXHOy dHqhuo"
    data-portal-slot="overlay"
  >
    <div
      class="sc-crXcEl fbCIbU"
    >
      <button
        aria-label="Close"
        class="sc-bczRLJ sc-kDDrLX eblJCO fWCYnN"
        type="button"
      >
        <svg
          aria-hidden="true"
          focusable="false"
          height="15px"
          version="1.1"
          viewBox="0 0 15 15"
          width="15px"
        >
          <g
            fill="none"
            fill-rule="evenodd"
            stroke="none"
            stroke-width="1"
          >
            <g
              fill="currentColor"
              transform="translate(-302.000000, -18.000000)"
            >
              <g
                transform="translate(302.000000, 18.000000)"
              >
                <path
                  d="M7.5,5.41522791 L12.0331524,0.579865364 C12.3077536,0.286957429 12.7165503,0.24816296 12.946282,0.493210121 L13.9861449,1.60239723 C14.2158766,1.84744439 14.1795068,2.28349422 13.9049056,2.57640216 L9.37175324,7.41176471 L13.9049056,12.2471273 C14.1795068,12.5400352 14.2158766,12.976085 13.9861449,13.2211322 L12.946282,14.3303193 C12.7165503,14.5753665 12.3077536,14.536572 12.0331524,14.243664 L7.5,9.4083015 L2.96684761,14.243664 C2.69224642,14.536572 2.2834497,14.5753665 2.05371799,14.3303193 L1.01385508,13.2211322 C0.784123363,12.976085 0.820493178,12.5400352 1.09509437,12.2471273 L5.62824676,7.41176471 L1.09509437,2.57640216 C0.820493178,2.28349422 0.784123363,1.84744439 1.01385508,1.60239723 L2.05371799,0.493210121 C2.2834497,0.24816296 2.69224642,0.286957429 2.96684761,0.579865364 L7.5,5.41522791 Z"
                />
              </g>
            </g>
          </g>
        </svg>
      </button>
      Inner content
    </div>
    <div
      class="sc-gKXOVf sc-jqUVSM etxkPo hjYSJA"
    />
  </div>
</body>
`;
